### PR TITLE
AddImportPlacementOptions

### PR DESCRIPTION
**Pull request by tmat**
Separate import placement options from CodeGenerationPreferences.

### CODE REVIEW

**Comment by tmat** [comment]
@CyrusNajmabadi ptal

File: src/EditorFeatures/Core/Implementation/AddImports/AbstractAddImportsPasteCommandHandler.cs [hunk 1]
```diff
@@ -127,7 +125,16 @@ public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, Com
...
+            var updatedDocument = _threadingContext.JoinableTaskFactory.Run(async () =>
+            {
+                var placement = await AddImportPlacementOptions.FromDocumentAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var options = new AddMissingImportsOptions(
+                    HideAdvancedMembers: document.Project.Solution.Options.GetOption(CompletionOptions.Metadata.HideAdvancedMembers, document.Project.Language),
+                    placement);
+
+                return await addMissingImportsService.AddMissingImportsAsync(document, textSpan, options, cancellationToken).ConfigureAwait(false);
+            });
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/AddImports/AbstractAddImportsPasteCommandHandler.cs:137]: note to @ryzngard we def want to make this totally async.

File: src/Workspaces/CSharpTest/CodeGeneration/AddImportsTests.cs [hunk 2]
```diff
@@ -193,7 +190,7 @@ class C
     public List<int> F;
 }",
                 useSymbolAnnotations,
-                options => options.WithChangedOption(GenerationOptions.PlaceSystemNamespaceFirst, LanguageNames.CSharp, false)
+                placeSystemNamespaceFirst: false
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/CodeGeneration/AddImportsTests.cs:193]: much nicer.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs [hunk 3]
```diff
@@ -29,6 +27,13 @@ public CSharpAddImportsService()
         {
         }
 
+#if !CODE_STYLE
+        public override bool PlaceImportsInsideNamespaces(Options.OptionSet optionSet)
+        {
+            return optionSet.GetOption(CSharpCodeStyleOptions.PreferredUsingDirectivePlacement).Value == AddImportPlacement.InsideNamespace;
+        }
+#endif
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs:35]: intereting.  how does this work with code style?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs [hunk 4]
```diff
@@ -203,18 +203,5 @@ public static IEnumerable<Document> GetLinkedDocuments(this Document document)
...
-        {
-#if !CODE_STYLE
-            // Normally we don't allow generation into a hidden region in the file.  However, if we have a
-            // modern span mapper at our disposal, we do allow it as that host span mapper can handle mapping
-            // our edit to their domain appropriate.
-            var spanMapper = document.Services.GetService<ISpanMappingService>();
-            return spanMapper != null && spanMapper.SupportsMappingImportDirectives;
-#else
-            return false;
-#endif
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:217]: did this logic move elsewhere?

**Review by CyrusNajmabadi — COMMENTED** [review]
overall looks good... but i'm mising a little how things work if you're in code_style.  I woulnd't have expected any strangeness there as the namespace-placement options should work for code-style as well.

Can you clarify?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs:35]: Not sure what do you mean, could you elaborate?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs:217]: Yes, moved to AddImportPlacementOptions record struct.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs:35]: yeah... i'm def curious what's going on here.  the layering seems strange.

Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpAddImportsService.cs:35]: Filed https://github.com/dotnet/roslyn/issues/59179

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**